### PR TITLE
Fixes #771: add/use interface for query validator

### LINKED ISSUE #771: Add/use interface for query validator
**Assemblies affected**
Which assemblies and versions are known to be affected e.g. ASP.NET Core OData 8.1

Introduce the interfaces for all odata query validator
use the interfaces to register the validators

### PR DESCRIPTION
**Pull request by xuzhg**
Add interfaces for query validator

This is one of step for issue #767

### CODE REVIEW

File: src/Microsoft.AspNetCore.OData/Query/Validator/Interfaces/IFilterQueryValidator.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    /// based on the <see cref="ODataValidationSettings"/>.
+    /// </summary>
+    public interface IFilterQueryValidator
+    {
+        /// <summary>
+        /// Validates the OData query.
+        /// </summary>
+        /// <param name="filterQueryOption">The $filter query.</param>
+        /// <param name="validationSettings">The validation settings.</param>
+        void Validate(FilterQueryOption filterQueryOption, ODataValidationSettings validationSettings);
```
Comment (corranrogue9) [inline src/Microsoft.AspNetCore.OData/Query/Validator/Interfaces/IFilterQueryValidator.cs:21]: This doesn't have a return type, so which is expected to happen if the query is invalid?

Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/Validator/Interfaces/IFilterQueryValidator.cs:21]: Good question. 
So far, we'd expect to throw exceptions if 'bad' happens.

But, I'd 're-think' it in the next step.

File: src/Microsoft.AspNetCore.OData/Query/Validator/Interfaces/IODataQueryValidator.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    /// <summary>
+    /// Provide the interface used to validate a <see cref="ODataQueryOptions"/>
+    /// based on the <see cref="ODataValidationSettings"/>.
+    /// </summary>
+    public interface IODataQueryValidator
+    {
+        /// <summary>
+        /// Validates the OData query.
+        /// </summary>
+        /// <param name="options">The OData query to validate.</param>
```
Comment (gathogojr) [inline src/Microsoft.AspNetCore.OData/Query/Validator/Interfaces/IODataQueryValidator.cs:19]:
```suggestion
        /// <param name="options">The OData query options to validate.</param>
```

**Review by gathogojr — DISMISSED** [review]
Left a few minor suggestions otherwise LGTM

File: src/Microsoft.AspNetCore.OData/Query/ODataQueryOptions.cs [hunk 3]
```diff
@@ -153,7 +153,7 @@ public ODataQueryOptions(ODataQueryContext context, HttpRequest request)
         /// <summary>
         /// Gets or sets the query validator.
         /// </summary>
-        public ODataQueryValidator Validator { get; set; }
+        public IODataQueryValidator Validator { get; set; }
```
Comment (mikepizzo) [inline src/Microsoft.AspNetCore.OData/Query/ODataQueryOptions.cs:156]: Isn't this a breaking change? Setting the property will still work, but any existing code that calls this property getter will expect an ODataQueryValidator, and will break if we change the return type to IODataQueryValidator.

File: src/Microsoft.AspNetCore.OData/Query/Query/ComputeQueryOption.cs [hunk 4]
```diff
@@ -113,7 +113,7 @@ public ComputeClause ComputeClause
         /// <summary>
         /// Gets or sets the $compute Query Validator.
         /// </summary>
-        public ComputeQueryValidator Validator { get; set; }
+        public IComputeQueryValidator Validator { get; set; }
```
Comment (mikepizzo) [inline src/Microsoft.AspNetCore.OData/Query/Query/ComputeQueryOption.cs:116]: Same as above; aren't all of these breaking changes?

File: src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs [hunk 5]
```diff
@@ -89,14 +89,14 @@ public virtual void Validate(OrderByQueryOption orderByOption, ODataValidationSe
             }
         }
 
-        internal static OrderByQueryValidator GetOrderByQueryValidator(ODataQueryContext context)
+        internal static IOrderByQueryValidator GetOrderByQueryValidator(ODataQueryContext context)
```
Comment (mikepizzo) [inline src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs:92]: Why would these not still return the concrete types?

**Review by mikepizzo — COMMENTED** [review]
@xuzhg  -- This PR looks like it introduces breaking changes. Let's discuss before we release.

File: src/Microsoft.AspNetCore.OData/Query/ODataQueryOptions.cs (hunk 3 above)
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/ODataQueryOptions.cs:156]: No, It's not a breaking change.  ODataQueryValidator implements the IODataQueryValidator interface.
The existing codes still can work.

File: src/Microsoft.AspNetCore.OData/Query/Query/ComputeQueryOption.cs (hunk 4 above)
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/Query/ComputeQueryOption.cs:116]: No, It's not a breaking change.  ComputeQueryValidator implements the IComputeQueryValidator interface.
The existing codes still can work.

File: src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs (hunk 5 above)
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs:92]: If we return concrete type, we can't let users extend/customize it.

**Pull request merged**